src/internal/document_api/mod.rs

```diff
@@ -915,6 +944,71 @@ pub fn decrypt_document<'a, CR: rand::CryptoRng + rand::RngCore>(
         })
 }
 
+/// Decrypt the unmanaged document. The caller must provide both the encrypted data as well as the
+/// encrypted DEKs. Most use cases would want `decrypt_document` instead.
+pub fn decrypt_document_unmanaged<'a, CR: rand::CryptoRng + rand::RngCore>(
+    auth: &'a RequestAuth,
+    recrypt: &'a Recrypt<Sha256, Ed25519, RandomBytes<CR>>,
+    device_private_key: &'a PrivateKey,
+    encrypted_doc: &'a [u8],
+    encrypted_deks: &'a [u8],
+) -> impl Future<Item = DocumentDecryptUnmanagedResult, Error = IronOxideErr> + 'a {
+    // attempt to parse the proto as fail-fast validation. If it fails decrypt will fail
+    protobuf::parse_from_bytes::<EncryptedDeksP>(encrypted_deks)
```

Comment:
Since we are going to pay the cost of proto de-serialization, we could also check that document_id and segment_id of the edoc and edeks match prior to making a network call as additional validation.

Reply:
I did this.